File: packages/sdk-node/tests/attachments.test.ts
import { beforeEach, describe, expect, it } from 'vitest';
import type { Bridge } from '../src/bridges/types';
import { MAX_ATTACHMENT_SIZE, PocketPing } from '../src/pocketping';
import type { Attachment, Message, Session, UploadRequest } from '../src/types';

/** Bridge that records the messages it receives. */
class RecordingBridge implements Bridge {
  name = 'recording';
  visitorMessages: Array<{ message: Message; session: Session }> = [];

  onVisitorMessage(message: Message, session: Session): void {
    this.visitorMessages.push({ message, session });
  }
}

function uploadReq(sessionId: string, overrides: Partial<UploadRequest> = {}): UploadRequest {
  return {
    sessionId,
    filename: 'photo.jpg',
    mimeType: 'image/jpeg',
    size: 1024,
    ...overrides,
  };
}

describe('File Attachments', () => {
  let pp: PocketPing;
  let sessionId: string;

  beforeEach(async () => {
    pp = new PocketPing();
    const conn = await pp.handleConnect({ visitorId: 'visitor-1' });
    sessionId = conn.sessionId;
  });

  // 1. Creates upload request with presigned URL
  it('creates upload request with presigned URL', async () => {
    const before = Date.now();
    const res = await pp.handleUploadRequest(uploadReq(sessionId));

    expect(res.attachmentId).toBeTruthy();
    expect(res.uploadUrl).toContain(res.attachmentId);
    expect(res.expiresAt.getTime()).toBeGreaterThan(before);

    const stored = await pp.getStorage().getAttachment?.(res.attachmentId);
    expect(stored?.status).toBe('pending');
    expect(stored?.messageId).toBeNull();
  });

  // 2. Marks attachment as ready after upload
  it('marks attachment as ready after upload', async () => {
    const res = await pp.handleUploadRequest(uploadReq(sessionId));

    const ready = await pp.handleUploadComplete(res.attachmentId);
    expect(ready.status).toBe('ready');

    const stored = await pp.getStorage().getAttachment?.(res.attachmentId);
    expect(stored?.status).toBe('ready');
  });

  // 3. Links attachments to message
  it('links attachments to message', async () => {
    const res = await pp.handleUploadRequest(uploadReq(sessionId));
    await pp.handleUploadComplete(res.attachmentId);

    const sent = await pp.handleMessage({
      sessionId,
      content: 'Here is a file',
      sender: 'visitor',
      attachmentIds: [res.attachmentId],
    });

    const stored = await pp.getStorage().getAttachment?.(res.attachmentId);
    expect(stored?.messageId).toBe(sent.messageId);
  });

  // 4. Returns attachments with message
  it('returns attachments with message', async () => {
    const res = await pp.handleUploadRequest(uploadReq(sessionId, { filename: 'doc.pdf', mimeType: 'application/pdf' }));
    await pp.handleUploadComplete(res.attachmentId);

    await pp.handleMessage({
      sessionId,
      content: 'doc attached',
      sender: 'visitor',
      attachmentIds: [res.attachmentId],
    });

    const { messages } = await pp.handleGetMessages({ sessionId });
    const withAttachment = messages.find((m) => m.attachments && m.attachments.length > 0);
    expect(withAttachment).toBeDefined();
    expect(withAttachment?.attachments?.[0].id).toBe(res.attachmentId);
    expect(withAttachment?.attachments?.[0].filename).toBe('doc.pdf');

    // Also returned via connect (resume)
    const reconnect = await pp.handleConnect({ visitorId: 'visitor-1', sessionId });
    const fromConnect = reconnect.messages.find((m) => m.attachments && m.attachments.length > 0);
    expect(fromConnect?.attachments?.[0].id).toBe(res.attachmentId);
  });

  // 5. Rejects invalid mime types
  it('rejects invalid mime types', async () => {
    await expect(
      pp.handleUploadRequest(uploadReq(sessionId, { mimeType: 'application/x-msdownload' }))
    ).rejects.toThrow();
  });

  // 6. Rejects files over size limit
  it('rejects files over size limit', async () => {
    await expect(
      pp.handleUploadRequest(uploadReq(sessionId, { size: MAX_ATTACHMENT_SIZE + 1 }))
    ).rejects.toThrow();
  });

  // 7. Handles upload failure gracefully
  it('handles upload failure gracefully', async () => {
    const res = await pp.handleUploadRequest(uploadReq(sessionId));

    const failed = await pp.handleUploadFailed(res.attachmentId);
    expect(failed.status).toBe('failed');

    // Unknown id returns null without crashing
    const unknown = await pp.getStorage().getAttachment?.('does-not-exist');
    expect(unknown ?? null).toBeNull();
  });

  // 8. Syncs attachments to bridges
  it('syncs attachments to bridges', async () => {
    const bridge = new RecordingBridge();
    const ppWithBridge = new PocketPing({ bridges: [bridge] });
    const conn = await ppWithBridge.handleConnect({ visitorId: 'visitor-2' });

    const res = await ppWithBridge.handleUploadRequest(uploadReq(conn.sessionId));
    await ppWithBridge.handleUploadComplete(res.attachmentId);

    await ppWithBridge.handleMessage({
      sessionId: conn.sessionId,
      content: 'with attachment',
      sender: 'visitor',
      attachmentIds: [res.attachmentId],
    });

    expect(bridge.visitorMessages.length).toBe(1);
    const received = bridge.visitorMessages[0].message;
    expect(received.attachments).toBeDefined();
    expect(received.attachments?.map((a: Attachment) => a.id)).toContain(res.attachmentId);
  });

  // Extra: validates session existence
  it('rejects upload requests for unknown sessions', async () => {
    await expect(pp.handleUploadRequest(uploadReq('unknown-session'))).rejects.toThrow(
      'Session not found'
    );
  });

  // Extra: respects overridable config
  it('respects overridable max size, mime list, and upload base url', async () => {
    const custom = new PocketPing({
      maxAttachmentSize: 100,
      allowedMimeTypes: ['image/png'],
      uploadBaseUrl: 'https://files.example.com',
    });
    const conn = await custom.handleConnect({ visitorId: 'visitor-3' });

    const res = await custom.handleUploadRequest(
      uploadReq(conn.sessionId, { mimeType: 'image/png', size: 50 })
    );
    expect(res.uploadUrl.startsWith('https://files.example.com/')).toBe(true);

    await expect(
      custom.handleUploadRequest(uploadReq(conn.sessionId, { mimeType: 'image/jpeg', size: 50 }))
    ).rejects.toThrow();
    await expect(
      custom.handleUploadRequest(uploadReq(conn.sessionId, { mimeType: 'image/png', size: 101 }))
    ).rejects.toThrow();
  });
});

File: packages/sdk-go/attachments_test.go
package pocketping

import (
	"context"
	"errors"
	"sync"
	"testing"
	"time"
)

// recordingBridge captures messages passed to OnVisitorMessage along with a
// snapshot of their attachments, in a concurrency-safe way.
type recordingBridge struct {
	BaseBridge
	mu          sync.Mutex
	messages    []Message
	attachments [][]Attachment
}

func newRecordingBridge(name string) *recordingBridge {
	return &recordingBridge{BaseBridge: BaseBridge{BridgeName: name}}
}

func (r *recordingBridge) OnVisitorMessage(ctx context.Context, message *Message, session *Session) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.messages = append(r.messages, *message)
	atts := make([]Attachment, len(message.Attachments))
	copy(atts, message.Attachments)
	r.attachments = append(r.attachments, atts)
	return nil
}

func (r *recordingBridge) lastAttachments() ([]Attachment, bool) {
	r.mu.Lock()
	defer r.mu.Unlock()
	if len(r.attachments) == 0 {
		return nil, false
	}
	return r.attachments[len(r.attachments)-1], true
}

// newSession creates a PocketPing instance and a session, returning both.
func newSessionFixture(t *testing.T, pp *PocketPing) string {
	t.Helper()
	ctx := context.Background()
	resp, err := pp.HandleConnect(ctx, ConnectRequest{VisitorID: "visitor-1"})
	if err != nil {
		t.Fatalf("connect failed: %v", err)
	}
	return resp.SessionID
}

// Scenario 1: Creates upload request with presigned URL.
func TestHandleUploadRequestCreatesPresignedURL(t *testing.T) {
	pp := New(Config{})
	ctx := context.Background()
	sessionID := newSessionFixture(t, pp)

	before := time.Now()
	resp, err := pp.HandleUploadRequest(ctx, UploadRequest{
		SessionID: sessionID,
		Filename:  "photo.jpg",
		MimeType:  "image/jpeg",
		Size:      1024,
	})
	if err != nil {
		t.Fatalf("expected no error, got %v", err)
	}

	if resp.AttachmentID == "" {
		t.Fatal("expected an attachment ID")
	}
	if !contains(resp.UploadURL, resp.AttachmentID) {
		t.Errorf("expected uploadUrl %q to contain attachmentId %q", resp.UploadURL, resp.AttachmentID)
	}
	if !resp.ExpiresAt.After(before) {
		t.Errorf("expected expiresAt %v to be in the future", resp.ExpiresAt)
	}

	// The stored attachment should be pending.
	att, err := pp.storage.(StorageWithAttachments).GetAttachment(ctx, resp.AttachmentID)
	if err != nil || att == nil {
		t.Fatalf("expected stored attachment, err=%v att=%v", err, att)
	}
	if att.Status != AttachmentStatusPending {
		t.Errorf("expected status pending, got %v", att.Status)
	}
}

// Scenario 2: Marks attachment as ready after upload.
func TestHandleUploadCompleteMarksReady(t *testing.T) {
	pp := New(Config{})
	ctx := context.Background()
	sessionID := newSessionFixture(t, pp)

	resp, err := pp.HandleUploadRequest(ctx, UploadRequest{
		SessionID: sessionID,
		Filename:  "doc.pdf",
		MimeType:  "application/pdf",
		Size:      2048,
	})
	if err != nil {
		t.Fatalf("upload request failed: %v", err)
	}

	att, err := pp.HandleUploadComplete(ctx, resp.AttachmentID)
	if err != nil {
		t.Fatalf("upload complete failed: %v", err)
	}
	if att.Status != AttachmentStatusReady {
		t.Errorf("expected status ready, got %v", att.Status)
	}

	// Verify persisted.
	stored, _ := pp.storage.(StorageWithAttachments).GetAttachment(ctx, resp.AttachmentID)
	if stored.Status != AttachmentStatusReady {
		t.Errorf("expected stored status ready, got %v", stored.Status)
	}
}

// Scenario 3: Links attachments to message.
func TestHandleMessageLinksAttachments(t *testing.T) {
	pp := New(Config{})
	ctx := context.Background()
	sessionID := newSessionFixture(t, pp)

	upload, err := pp.HandleUploadRequest(ctx, UploadRequest{
		SessionID: sessionID,
		Filename:  "report.csv",
		MimeType:  "text/csv",
		Size:      512,
	})
	if err != nil {
		t.Fatalf("upload request failed: %v", err)
	}

	msg, err := pp.HandleMessage(ctx, SendMessageRequest{
		SessionID:     sessionID,
		Content:       "here is the report",
		Sender:        SenderVisitor,
		AttachmentIDs: []string{upload.AttachmentID},
	})
	if err != nil {
		t.Fatalf("handle message failed: %v", err)
	}

	att, _ := pp.storage.(StorageWithAttachments).GetAttachment(ctx, upload.AttachmentID)
	if att == nil {
		t.Fatal("expected attachment to exist")
	}
	if att.MessageID != msg.MessageID {
		t.Errorf("expected attachment.messageId=%q, got %q", msg.MessageID, att.MessageID)
	}
}

// Scenario 4: Returns attachments with message (get messages / connect).
func TestGetMessagesReturnsAttachments(t *testing.T) {
	pp := New(Config{})
	ctx := context.Background()
	sessionID := newSessionFixture(t, pp)

	upload, _ := pp.HandleUploadRequest(ctx, UploadRequest{
		SessionID: sessionID,
		Filename:  "image.png",
		MimeType:  "image/png",
		Size:      4096,
	})
	if _, err := pp.HandleMessage(ctx, SendMessageRequest{
		SessionID:     sessionID,
		Content:       "see attached",
		Sender:        SenderVisitor,
		AttachmentIDs: []string{upload.AttachmentID},
	}); err != nil {
		t.Fatalf("handle message failed: %v", err)
	}

	// Via HandleGetMessages.
	got, err := pp.HandleGetMessages(ctx, GetMessagesRequest{SessionID: sessionID})
	if err != nil {
		t.Fatalf("get messages failed: %v", err)
	}
	if len(got.Messages) != 1 {
		t.Fatalf("expected 1 message, got %d", len(got.Messages))
	}
	if len(got.Messages[0].Attachments) != 1 {
		t.Fatalf("expected 1 attachment on message, got %d", len(got.Messages[0].Attachments))
	}
	if got.Messages[0].Attachments[0].ID != upload.AttachmentID {
		t.Errorf("expected attachment id %q, got %q", upload.AttachmentID, got.Messages[0].Attachments[0].ID)
	}

	// Via HandleConnect (resume session).
	conn, err := pp.HandleConnect(ctx, ConnectRequest{VisitorID: "visitor-1", SessionID: sessionID})
	if err != nil {
		t.Fatalf("connect failed: %v", err)
	}
	if len(conn.Messages) != 1 || len(conn.Messages[0].Attachments) != 1 {
		t.Fatalf("expected connect to return message with 1 attachment, got %+v", conn.Messages)
	}
}

// Scenario 5: Rejects invalid mime types.
func TestHandleUploadRequestRejectsInvalidMime(t *testing.T) {
	pp := New(Config{})
	ctx := context.Background()
	sessionID := newSessionFixture(t, pp)

	_, err := pp.HandleUploadRequest(ctx, UploadRequest{
		SessionID: sessionID,
		Filename:  "evil.exe",
		MimeType:  "application/x-msdownload",
		Size:      1024,
	})
	if !errors.Is(err, ErrInvalidMimeType) {
		t.Errorf("expected ErrInvalidMimeType, got %v", err)
	}
}

// Scenario 6: Rejects files over size limit.
func TestHandleUploadRequestRejectsOversizeFile(t *testing.T) {
	pp := New(Config{})
	ctx := context.Background()
	sessionID := newSessionFixture(t, pp)

	_, err := pp.HandleUploadRequest(ctx, UploadRequest{
		SessionID: sessionID,
		Filename:  "huge.pdf",
		MimeType:  "application/pdf",
		Size:      DefaultMaxAttachmentSize + 1,
	})
	if !errors.Is(err, ErrFileTooLarge) {
		t.Errorf("expected ErrFileTooLarge, got %v", err)
	}

	// Zero/negative sizes are rejected too.
	if _, err := pp.HandleUploadRequest(ctx, UploadRequest{
		SessionID: sessionID,
		Filename:  "empty.pdf",
		MimeType:  "application/pdf",
		Size:      0,
	}); !errors.Is(err, ErrFileTooLarge) {
		t.Errorf("expected ErrFileTooLarge for zero size, got %v", err)
	}
}

// Scenario 7: Handles upload failure gracefully.
func TestHandleUploadFailedAndUnknownAttachment(t *testing.T) {
	pp := New(Config{})
	ctx := context.Background()
	sessionID := newSessionFixture(t, pp)

	upload, _ := pp.HandleUploadRequest(ctx, UploadRequest{
		SessionID: sessionID,
		Filename:  "broken.zip",
		MimeType:  "application/zip",
		Size:      1024,
	})

	att, err := pp.HandleUploadFailed(ctx, upload.AttachmentID)
	if err != nil {
		t.Fatalf("upload failed handler errored: %v", err)
	}
	if att.Status != AttachmentStatusFailed {
		t.Errorf("expected status failed, got %v", att.Status)
	}

	// Unknown attachment id returns nil without crashing in storage,
	// and ErrAttachmentNotFound from the handlers.
	missing, gerr := pp.storage.(StorageWithAttachments).GetAttachment(ctx, "does-not-exist")
	if gerr != nil || missing != nil {
		t.Errorf("expected (nil, nil) for unknown attachment, got (%v, %v)", missing, gerr)
	}
	if _, herr := pp.HandleUploadComplete(ctx, "does-not-exist"); !errors.Is(herr, ErrAttachmentNotFound) {
		t.Errorf("expected ErrAttachmentNotFound, got %v", herr)
	}
}

// Scenario 8: Syncs attachments to bridges.
func TestSyncsAttachmentsToBridges(t *testing.T) {
	bridge := newRecordingBridge("recorder")
	pp := New(Config{Bridges: []Bridge{bridge}})
	ctx := context.Background()
	sessionID := newSessionFixture(t, pp)

	upload, _ := pp.HandleUploadRequest(ctx, UploadRequest{
		SessionID: sessionID,
		Filename:  "screenshot.png",
		MimeType:  "image/png",
		Size:      8192,
	})

	if _, err := pp.HandleMessage(ctx, SendMessageRequest{
		SessionID:     sessionID,
		Content:       "look at this",
		Sender:        SenderVisitor,
		AttachmentIDs: []string{upload.AttachmentID},
	}); err != nil {
		t.Fatalf("handle message failed: %v", err)
	}

	// Bridge notification is async; wait for it to be recorded.
	var atts []Attachment
	deadline := time.Now().Add(2 * time.Second)
	for time.Now().Before(deadline) {
		if a, ok := bridge.lastAttachments(); ok {
			atts = a
			break
		}
		time.Sleep(5 * time.Millisecond)
	}

	if len(atts) != 1 {
		t.Fatalf("expected bridge to receive 1 attachment, got %d", len(atts))
	}
	if atts[0].ID != upload.AttachmentID {
		t.Errorf("expected bridge attachment id %q, got %q", upload.AttachmentID, atts[0].ID)
	}
	if atts[0].MessageID == "" {
		t.Error("expected bridge attachment to have a linked messageId")
	}
}

// Bonus: session-not-found is surfaced from HandleUploadRequest.
func TestHandleUploadRequestSessionNotFound(t *testing.T) {
	pp := New(Config{})
	ctx := context.Background()

	_, err := pp.HandleUploadRequest(ctx, UploadRequest{
		SessionID: "nope",
		Filename:  "x.png",
		MimeType:  "image/png",
		Size:      1024,
	})
	if !errors.Is(err, ErrSessionNotFound) {
		t.Errorf("expected ErrSessionNotFound, got %v", err)
	}
}

// contains is a tiny substring helper to avoid importing strings in tests.
func contains(haystack, needle string) bool {
	if needle == "" {
		return true
	}
	for i := 0; i+len(needle) <= len(haystack); i++ {
		if haystack[i:i+len(needle)] == needle {
			return true
		}
	}
	return false
}
